src/diff/mod.rs

```diff
@@ -1,103 +1,174 @@
 }
 
-pub fn kdbx_to_sorted_vec(
-  file: &str,
-  password: Option<String>,
-  keyfile_path: Option<&str>,
-) -> Result<SortedKdbxEntries> {
-  let mut keyfile = keyfile_path.map(|path| File::open(Path::new(path)).unwrap());
-  File::open(Path::new(file))
-    .map_err(|e| Error::from(e))
-    .and_then(|mut db_file| {
-      let db = Database::open(
-        &mut db_file,
-        password.as_ref().map(|s| s.as_str()),
-        keyfile.as_mut().map(|f| f as &mut dyn Read),
-      );
-      db
-    })
-    .map(|db: Database| accumulate_all_entries(db.root))
+/// Helper wrapper to impl Display for a DiffResult with user-specified settings
+pub struct DiffDisplay<T: DiffResultFormat> {
+    pub inner: T,
+    pub depth: usize,
+    pub use_color: bool,
+}
+
+impl<T: DiffResultFormat> std::fmt::Display for DiffDisplay<T> {
+    fn fmt(&self, mut f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        self.inner
+            .diff_result_format(&mut f, self.depth, self.use_color)
+    }
 }
 
-fn accumulate_all_entries(start: Group) -> SortedKdbxEntries {
-  let mut accumulated = check_group(&mut Vec::new(), &mut Vec::new(), start);
-  accumulated.sort();
-  accumulated.dedup();
-  accumulated
+/// Format functionality for deep recursion
+impl<'a, E> DiffResultFormat for DiffResult<'a, E>
+where
+    E: std::fmt::Display + std::fmt::Debug,
+{
+    fn diff_result_format(
+        &self,
+        mut f: &mut std::fmt::Formatter<'_>,
+        depth: usize,
+        use_color: bool,
+    ) -> std::fmt::Result {
+        let indent = "  ".repeat(depth);
+        match self {
+            DiffResult::Identical { .. } => Ok(()),
+            DiffResult::Changed { left, right } => {
+                if use_color {
+                    crate::set_fg(Some(Color::Red));
+                }
+                write!(f, "{}- {}\n", indent, left)?;
+                if use_color {
+                    crate::set_fg(Some(Color::Green));
+                }
+                write!(f, "{}+ {}\n", indent, right)
+            }
+            DiffResult::InnerDifferences {
+                left,
+                inner_differences,
+                ..
+            } => {
+                if use_color {
+                    crate::set_fg(Some(Color::Yellow));
+                }
+                write!(f, "{}~ {}\n", indent, left)?;
+                for id in inner_differences {
+                    id.diff_result_format(&mut f, depth + 1, use_color)?;
+                }
+                Ok(())
+            }
+            DiffResult::OnlyLeft { left } => {
+                if use_color {
+                    crate::set_fg(Some(Color::Red));
+                }
+                write!(f, "{}- {}\n", indent, left)
+            }
+            DiffResult::OnlyRight { right } => {
+                if use_color {
+                    crate::set_fg(Some(Color::Green));
+                }
+                write!(f, "{}+ {}\n", indent, right)
+            }
+        }?;
```

Comment:
The [? operator](https://doc.rust-lang.org/edition-guide/rust-2018/error-handling-and-panics/the-question-mark-operator-for-easier-error-handling.html) is the way to get the value within a `Result` type in a properly-error-handled way.

In this case, all of the match cases are returning a `std::fmt::Result` from the `write!` operation. If the operation succeeds with a return value of `Ok(())`, this gets ignored. If there is an `Err(some_error)`, this will get returned for the whole `diff_result_format` call's return value.

Reply:
Probably it would also make sense in this case to silently ignore any write errors and still attempt to always send the color reset below. Would you like me to adjust this?

Reply:
I added another commit to do so.